Implement `RubyUtils.find_executable`.

noniq opened the pull request:
First try on implementing #102. Not sure about the module name `RubyUtils`, though – happy to take suggestions for alternative names!

TODO: Do we need to do something special to support RVM?

sorbits commented:
@jacob-carlborg, @sanssecours Any input on this one?

(my own knowledge of the ruby “tool chain” and the multitude of ways to execute it is rather limited)

jacob-carlborg commented:
I'll take a look. Is intended to be used by other bundles, in that case how?

noniq commented:
I plan on using it in the RSpec and RuboCop bundles. Both are now implementing a quite similar logic on their own. Would also make sense to use it for #99 for example. (I think quite a large part of that issue’s discussion could have been avoided if such a feature had already been present :-)

jacob-carlborg commented:
Well, how do you actually require a file from another bundle?

noniq commented:
Allan mentioned this once on the mailing list, its also described here: http://manual.textmate.org/bundles.html#requirements :-)

jacob-carlborg commented:
Ah, I see.

jacob-carlborg commented on Support/lib/ruby_utils.rb:
```diff
@@ -0,0 +1,55 @@
+  #
+  # Supports the following cases:
+  #
+  # 1. If an appropriate `TM_*` environment variable is present and points to an
```
I move this to the third step.

noniq replied:
It started there initially :) But then I realized that the env var is the most “finegrained” setting: I can always add a `.tm_properties` file to a project (or even a subdir) setting this variable to override whatever else setting would take effect normally.

A (slightly contrived) example: I have a project where the Gemfile contains `rubocop`, but for some reason I want to override this and use another version of RuboCop when running it from TextMate. If the env var has the highest priority, I can do so by simply setting `TM_RUBOCOP` in this project’s `.tm_properties` file.

But to be honest, I’m totally not sure about this 🤔 /cc @sorbits

jacob-carlborg replied:
I have the opposite view. I see the environment variable as a global fallback for all projects, that will be used if the Gemfile doesn't contain the requested gem.

sorbits replied:
> But then I realized that the env var is the most “finegrained” setting

I agree, the environment variable is there for when the default behavior is undesured and require user intervention, so it should trump everything else.

noniq replied:
Both views totally make sense IMHO, that’s the problem … With the env var having highest priority the “global fallback” use case is possible, too, at least in princinple: Set the env var globally to the fallback value, and explicitly unset it (set it to an empty string) for projects where you don’t want this global value to be used. But that’s feels rather awkward …

jacob-carlborg replied:
This is what I see will happen. An environment variable is configured to run a command for a project/simple script that does not have a Gemfile. Later when running the same command for a project that do have a Gemfile with the requested gem, it will not be used because the environment variable i used instead. I think it would be very easy to forget that the environment variable was set, which then would potentially cause a lot of pain to debug because the wrong version of a gem is used.

Also, to use this for both projects with a Gemfile and without one, I would need to constantly activate and disable the environment variable, if I want the gem from the Gemfile to have higher precedence.

jacob-carlborg replied:
> I can always add a .tm_properties file to a project 

If you would add a .tm_properties file, why not a Gemfile?

jacob-carlborg replied:
I think it would be more rare to override what's in the Gemfile than using the command for a file/project without a Gemfile.

sorbits replied:
> Also, to use this for both projects with a Gemfile and without one, I would need to constantly activate and disable the environment variable

We can add a “project scope” for projects with a `Gemfile` so that we can target these (or target projects without), here’s the list of current project attributes: https://github.com/textmate/textmate/blob/master/Frameworks/DocumentWindow/src/DocumentController.mm#L1429

It is true that users are not unlikely to set (globally) environment variables and forget about them, and then later something breaks because of this, but I think that goes for pretty much all environment variables.

Not having it trump everything means that if a user reports that for some project, our logic is wrong, we cannot provide them with any workaround, they’ll have to wait for someone to fix the code and push a new release, that’s IMO where the environment variable comes in: to allow users to fix things locally.

Using it as a fallback, IMHO that is the job of the `PATH` variable. I.e. ensure that whatever fallback executable you want to use when there is no `TM_«executable»`, no `Gemfile`, no etc., can be found via `PATH`.

jacob-carlborg replied:
> We can add a “project scope” for projects with a Gemfile so that we can target these (or target projects without), here’s the list of current project attributes: https://github.com/textmate/textmate/blob/master/Frameworks/DocumentWindow/src/DocumentController.mm#L1429

How would that work?

> Not having it trump everything means that if a user reports that for some project, our logic is wrong, we cannot provide them with any workaround, they’ll have to wait for someone to fix the code and push a new release, that’s IMO where the environment variable comes in: to allow users to fix things locally.

Not sure I understand. By default it would use what's in the Gemfile, the user has already control of the Gemfile.

jacob-carlborg commented:
> TODO: Do we need to do something special to support RVM?

Yes, something. I'm not sure what. It's easy to do from a shell script but I'm not sure how to do it from within Ruby.

sorbits commented:
> > We can add a “project scope” for projects with a Gemfile […]
> How would that work?

In `.tm_properties` you can presently do this:

	FOO = 'no-make'

	[ attr.project.make ]
	FOO = "make"

Now try `echo $FOO` in a document in a project with a `Makefile` and one without, and you’ll see different values.

The primary use of this is key bindings, e.g. <kbd>⌘Y</kbd> shows the proper SCM actions and <kbd>⌘B</kbd> invokes the proper build command…

> > Not having it trump everything means that if a user reports that for some project, our logic is wrong, we cannot provide them with any workaround
> Not sure I understand. By default it would use what's in the Gemfile, the user has already control of the Gemfile.

And the `Gemfile` is always correct?

For example is it unthinkable that user has a project being run both in Vagrant and on macOS, the `Gemfile.lock` file would be shared between the two, but something could prevent it from executing w/o issues on the Mac?

As I said initially, my knowledge of the ruby “tool chain” is very limited, but I think it’s standard practice to have the environment variable trump everything.

So I’d rather ask you, why do you set an environment variable for a fallback option instead of adding the location of the fallback executable to `PATH`?

noniq commented:
> Using it as a fallback, IMHO that is the job of the PATH variable. I.e. ensure that whatever fallback executable you want to use when there is no TM_«executable», no Gemfile, no etc., can be found via PATH.

I think that’s a compelling argument. And it’s simple to explain, too: “Setup PATH so that it includes all the executables you want to use by default, use TM_«foo» for project-specific overrides.”

> If you would add a .tm_properties file, why not a Gemfile?

In shared projects it may not be possible to change the Gemfile just because of personal preferences / configuration choices. Adding a `.tm_properties` file (either Git-ignored or one directory above the Git root) seems more suitable.

jacob-carlborg commented:
> In .tm_properties you can presently do this:
FOO = 'no-make'
[ attr.project.make ]
FOO = "make"
Now try echo $FOO in a document in a project with a Makefile and one without, and you’ll see different values.

Hmm, then I would need to set something like `TM_RUBOCOP = 'bundle exec rubocop'`, which is not compatible with the current code in this PR, `File.executable?(ENV[env_var])`. I'm also not sure if the system will interpret the whole `bundle exec rubocop` as one single command or as one command with two arguments.

> For example is it unthinkable that user has a project being run both in Vagrant and on macOS, the Gemfile.lock file would be shared between the two, but something could prevent it from executing w/o issues on the Mac?

That's a possibility. I would say that the Gemfile is broken in that case.

> So I’d rather ask you, why do you set an environment variable for a fallback option instead of adding the location of the fallback executable to PATH?

I don't know. Maybe because the documentation for most bundles say that the `TM_...` environment variable should be set.

jacob-carlborg commented:
I guess I could try adding the necessary executables to `PATH` instead.

jacob-carlborg commented:
Would it make sense to make an exectuable script that can be used from other languages than Ruby?

sanssecours commented:
> Not sure about the module name RubyUtils, though – happy to take suggestions for alternative names!

How about something like `Executable.find` or `RubyCommand.find` instead of `RubyUtils.find_executable`?

>> TODO: Do we need to do something special to support RVM?

> Yes, something. I'm not sure what. It's easy to do from a shell script but I'm not sure how to do it from within Ruby.

In a shell script you usually just source `"$HOME/.rvm/scripts/rvm"` to support RVM. I did not find anything about supporting RVM inside a Ruby script though. One **super ugly hack** to add this support: Run a Bash script inside the Ruby script. For example, the following command prints the path to the RVM ruby executable:

```rb
puts `bash -c '[[ -f "$HOME/.rvm/scripts/rvm" ]]; \
                 . "$HOME/.rvm/scripts/rvm"; 
               which ruby'`
```

By the way: I ported “Reformat Document” (Issue #99) to Ruby on my machine:

```rb
#!usr/bin/ruby

require 'shellwords'
require ENV['TM_BUNDLE_SUPPORT'] + '/lib/ruby_utils'
require ENV['TM_SUPPORT_PATH'] + '/lib/tm/detach'

Dir.chdir(ENV['TM_PROJECT_DIRECTORY'] || File.dirname(ENV['TM_FILEPATH'].to_s))

rubocop = RubyUtils.find_executable('rubocop')
unless rubocop
  error_message = 'This command requires RuboCop. ' \
                  'Please install RuboCop via “gem install rubocop”.'
  `"$DIALOG" tooltip --text "#{error_message}"`
end
rubocop = rubocop.join ' '

aha = `which aha`.rstrip
if aha
  color = 'n'
  output_format = '--html'
else
  output_format = '--text'
end

TextMate.detach do
  output = `#{rubocop} -a#{color.to_s} "$TM_FILEPATH" #{'| aha' if aha}`
  `"$DIALOG" tooltip #{output_format} #{output.shellescape}`
end
```

It works for my purposes, since I only use `rbenv` (and neither RVM or Bundler). I also tested the Bundler support which works too.

> Hmm, then I would need to set something like `TM_RUBOCOP = 'bundle exec rubocop'`, which is not compatible with the current code in this PR, `File.executable?(ENV[env_var])`.

Looks like we should remove the check above, especially since `RubyUtils.find_executable('rubocop')` might return something like `['bundle', 'exec', 'rubocop']`, if the current project contains a `Gemfile`.

sorbits commented:
I like the simpler `Executable.find` API.

As for `TM_RUBOCOP = 'bundle exec rubocop'` I think this should be supported.

So I guess the check needs to be a little more complex, e.g.:

    def is_shell_executable?(command, search = ENV['PATH'].to_s)
      if exe = command.to_s.shellsplit.first
        search.split(':').any? do |dir|
          File.executable?(File.expand_path(exe, dir))
        end
      else
        false
      end
    end

Or should we just assume that if the variable is set, then it should be used, as ignoring misconfigured values might be doing the user a disservice.

jacob-carlborg commented:
> Or should we just assume that if the variable is set, then it should be used, as ignoring misconfigured values might be doing the user a disservice.

I think so.

noniq commented:
`Executable.find` sounds nice. Does not refer specifically to Ruby executables, though – not sure if this could lead to misunderstandings? (On the other hand, apart from the Gemfile part, the functionality is Ruby-agnostic anyway [which, on the other other hand, may lead to the question if TextMate should have a generic `find_executable` API in bundle support?])

Anyway, I’m fine with both `Executable.find` and `RubyExecutable.find`.

I’ll update the code to support space-separated values in the TM environment vars and to do no explicit checking of the env var values. And I’ll look deeper into RVM and try to come up with a solution.

sorbits commented:
For general API: What other things could you see using this?

TextMate already has built-in support for requiredCommands, so if there 
is something re-usable about this ruby stuff, then probably we should 
rather update TextMate’s native support (since that does provide a 
better user experience).

As this API is in the Ruby bundle, I think that it’s “only for” 
Ruby is implied.

On 30 Aug 2016, at 18:08, Stefan Daschek wrote:

> `Executable.find` sounds nice. Does not refer specifically to Ruby 
> executables, though – not sure if this could lead to 
> misunderstandings? (On the other hand, apart from the Gemfile part, 
> the functionality is Ruby-agnostic anyway [which, on the other other 
> hand, may lead to the question if TextMate should have a generic 
> `find_executable` API in bundle support?])
>
> Anyway, I’m fine with both `Executable.find` and 
> `RubyExecutable.find`.
>
> I’ll update the code to support space-separated values in the TM 
> environment vars and to do no explicit checking of the env var values. 
> And I’ll look deeper into RVM and try to come up with a solution.
>
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/textmate/ruby.tmbundle/pull/104#issuecomment-243491951

sanssecours commented:
> …as ignoring misconfigured values might be doing the user a disservice.

A simple check – like the one Allan wrote – for the correct value of an environment variable such as `TM_RUBOCOP` would be nice. I think in the case of an incorrect value `RubyUtils.find_executable` could  just show a window via `"$DIALOG" alert` – containing an error message about the incorrect value of the environment variable – and then exit. This way, the user of the API does not  need to care about incorrect values of environment variables.

It might even make sense to guarantee that if `RubyUtils.find_executable` returns a command (and not `nil`), that this command is executable. After all, the “executable” is right there in the name.

noniq commented:
> I think in the case of an incorrect value RubyUtils.find_executable could just show a window via "$DIALOG" alert – containing an error message about the incorrect value of the environment variable – and then exit.

I think handling the “executable not found” case belongs to the caller. For example, in the RuboCop bundle I want to do handle this differently based on whether the command was invoked implicitly via did-save callback (show an unintrusive tooltip) or explicitly (show an alert with detailled error message).

> It might even make sense to guarantee that if RubyUtils.find_executable returns a command (and not nil), that this command is executable. After all, the “executable” is right there in the name.

That was my original intention. I’m not sure how to do this though if we consider cases like `bundle exec foobar`. Especially I do not want to really invoke the command only for testing if it’s present and executable – some commands might need an non-negligible startup time (for example `bundle exec rspec` in a Rails project with a big Gemfile).

sanssecours commented:
> I think handling the “executable not found” case belongs to the caller.

You are right, but then it would be nice if the function could provide more information about the error case. Otherwise the caller has to – more or less – do the same checks as the function, if he wants to provide the user of the TextMate command with a helpful error message. I think it would be best if the command would just raise different exceptions in the error case. Then the caller could even use the message of the exception as default error message for a TextMate command.

> That was my original intention. I’m not sure how to do this though if we consider cases like `bundle exec foobar`.

I think the simple check which just looks at the first part of the command – i.e. `File.executable?('bundle')` in the case of `bundle exec foobar` – is enough. After all, if `bundle` is executable, then so is `bundle exec foobar`.

noniq commented:
> I think it would be best if the command would just raise different exceptions in the error case.

Hm, which different exceptions could that be (apart from “executable not found“)?

> I think the simple check which just looks at the first part of the command – i.e. `File.executable?('bundle')` in the case of `bundle exec foobar` – is enough. After all, if `bundle` is executable, then so is `bundle exec foobar`.

Not sure: `bundle exec foobar` could still fail if `foobar` is not installed. Remember that this is the codepath where we’re looking at `ENV["TM_FOOBAR"]`, so we should not make any assumptions about Gemfiles etc. (There is a similar problem with rbenv: Rbenv shims are always in the search path, even if a gem isn’t installed for the current Ruby version. That’s why I implemented [this additional check](https://github.com/die-antwort/ruby.tmbundle/blob/3c5888a6a7a2367752955f2df7efe8cee7d1efe1/Support/lib/ruby_utils.rb#L51). But it works for the specific case of rbenv shims only.)

jacob-carlborg commented:
I think that the method not should try to be too cleaver, that will most likely cause issues because it's used in some unexpected way. If `ENV["TM_FOOBAR"]` is set, just execute that, it's the users that sets it anyway.

sanssecours commented:
> Hm, which different exceptions could that be (apart from “executable not found“)?

Something like `EnvironmentVariableNotExecutableError` together with an error message might be nice. For example, if the value of the environment variable is not executable we could raise `EnvironmentVariableNotExecutableError` providing an error message such as:

```
The value of the environment variable ”TM_RUBOCOP“: 

     “does_not_exist/rubocop”

does not specify a valid executable command.
```

> Not sure: `bundle exec foobar` could still fail if `foobar` is not installed.

That is of course true, but calling some command can always fail. If we use the simple check for executability, then the caller can always be sure that he is able to call the returned command. After running the command, he can then check the exit code to verify that everything went fine.

> There is a similar problem with rbenv: Rbenv shims are always in the search path, even if a gem isn’t installed for the current Ruby version. That’s why I implemented this additional check. But it works for the specific case of rbenv shims only.

I think the more checks the API provides, the nicer it is to use. Of course we can not check everything before we run the command, but basic executability should be doable.

noniq commented:
I like the idea of providing more info about the cause if the executable wasn’t found, so the method now raises an error with a (hopefully) helpful error message in these cases.

Also renamed the method to `Executable.find`, fixed a bug where env var settings would not work for executables in PATH, and made sure that env var settings like `bundle exec foobar` work correctly.

Still to do: Investigate RVM.

jacob-carlborg commented on Support/lib/executable.rb:
```diff
@@ -0,0 +1,72 @@
+    # shell escaping later on
+    raise ArgumentError, "Invalid characters found in '#{name}'" unless name =~ /\A[\w_-]+\z/
+
+    env_var ||= 'TM_' + name.gsub(/\W+/, '_').upcase
```
Isn't the `/\W+/` part covered by the line above?

noniq replied:
We allow `name` to contain dashes, and I think we have to replace them with underscores when deriving the name of the environment variable. But maybe there’s a simpler way, I’ll think about it!

jacob-carlborg commented on Support/lib/executable.rb:
```diff
@@ -0,0 +1,72 @@
+    env_var ||= 'TM_' + name.gsub(/\W+/, '_').upcase
+    if (cmd = ENV[env_var]) && cmd != ''
+      cmd = cmd.shellsplit
+      if system('which', '-s', cmd[0])
```
I don't like that the function is trying to be this clever. The user of this function would need to handle errors when invoking the executable anyway.

noniq replied:
The advantage of doing this check here is that we can provide an error message telling the user that there seems to be a problem with an environment variable specifically.

Imagine the environment variable is set to something that does no (longer) exist. If we omit this check, then command execution will fail later on with a generic error message, and the user may wonder “why the heck is this trying to invoke `foobar`?” before realizing that there is an env var setting he/she had already forgotten about.

So I think in this case the added benefit makes up for the added complexity of the code.

jacob-carlborg commented on Support/lib/executable.rb:
```diff
@@ -0,0 +1,72 @@
+    elsif File.exist?("bin/#{name}")
+      ["bin/#{name}"]
+
+    elsif File.exist?('Gemfile.lock') && File.read('Gemfile.lock') =~ /^    #{name} /
```
Is it better to execute `bundle show #{name}` and check the exist status?

noniq replied:
Sounds like a good idea! I’m a bit concerned about startup time overhead, though: When I run `bundle show rails` in an average Rails project on my machine, it takes almost 0.7 seconds to complete. (This is with bundler 1.12.5 on Ruby 2.1.6, will investigate other Ruby versions later.)

jacob-carlborg replied:
True, but reading Gemfile.lock doesn't feel very reliable. At least it shouldn't be a fixed number of spaces in the regular expression.

jacob-carlborg replied:
Yeah, just doing `cat` and `grep` is significantly faster.

jacob-carlborg replied:
BTW, there's no way it's possible to run `bundle exec ...` without having a lock file?

noniq replied:
Reading `Gemfile.lock` is indeed a bit hacky, but I think it’s justifiable. The lockfile format is fairly stable. We need to check for exactly 4 spaces, because these are the lines listing the bundled gems (lines with 6 spaces list dependencies). A lockfile consists of “sections”, so strictly speaking we should restrict our search to lines from the `GEM` section. But I don’t think this is necessary. The chance of getting a false positive is negligible: Currently, the `GEM` section is the only one containing lines starting with 4 (and more) spaces.

It would be nice to use `Bundler::LockfileParser.new(File.read('Gemfile.lock'))`. However, this requires that the Ruby version used for running this script has bundler installed – and unfortunately that seems not to be the case for the system ruby coming with Mac OS (tested with both the `ruby18` and `ruby20` shims). So I don’t think this is an option 😞

jacob-carlborg replied:
Ok.

jacob-carlborg commented:
@noniq the Ruby on Rails bundle has a "Set RVM..." command, that might be interesting to take a look at.

jacob-carlborg commented:
@noniq I might have a solution for RVM, the `do` action of the `rvm` command. From the documentation:

```
## Usage
    $ rvm [all|all-gemsets|<ruby>,...|<path>] [--verbose|--summary|--yaml|--json] do <command> ...
    $ rvm in <path> do <some-command> ...
Executes arbitrary commands against given a set of rvm environments
```

So what you can do is something like this `rvm #{ENV['TM_PROJECT_DIRECTORY'} do <command>`. To execute the command in the given directory, prefix the path with `in`, like this:

```
rvm in #{ENV['TM_PROJECT_DIRECTORY']} do <command>
```

noniq commented:
Great, thanks for investigating!

I wonder: Should we test for the existence of `~/.rvm/bin/rmv` and use this, or should we check if `rvm` is in the path (assuming that someone who wants RVM to be used has set up their $PATH in TextMate accordingly)?

jacob-carlborg commented:
We need to test if `~/.rvm/bin/rvm` exits. If RVM is setup properly it will not be in the PATH, because the `rvm` command will be a Bash function and not an executable. And it needs to have the RVM file in the project directory, if not this error will occur:

```
Could not determine which Ruby to use; . should contain .rvmrc or .versions.conf or
.ruby-version or .rbfu-version or .rbenv-version, or an appropriate line in Gemfile.
( see: 'rvm usage' )
```

noniq commented:
As far as I understand (see https://rvm.io/workflow/scripting), RVM is both an exectuable in path and a shell function. 

I was thinking along the lines of “If you want to use something non-standard from within TextMate, make sure to setup TextMate’s `$PATH` accordingly.” (And it would simplify the code a bit becaue we wouldn’t need to check for both a user and a system wide install of RVM). But not sure how common that thinking is, that’s why I asked on the mailing list, to hopefully get input from people actually using RVM :-)

Apart from this minor detail, I think what the code then needs to do is:
 * Use `rvm do . which «executable»` to check if the executable was installed for the given Ruby version.
 * If yes, return `%w(rvm do . #{executable}` as invocation to run this executable.

Did I miss something?

jacob-carlborg commented:
> As far as I understand (see https://rvm.io/workflow/scripting), RVM is both an exectuable in path and a shell function.

Yes, but the binary doesn't have all the features as the function. When RVM is used from the command line and it's always used as a function.

When RVM is used from within TextMate, the `TM_RUBY` variable should be set to `~/.rvm/bin/rvm-auto-ruby` which will automatically switch to the correct version based on the current directory [1]. At least I have not had a reason to add RVM to the PATH, because the `rvm` command itself it not invoked, it's the `rvm-auto-ruby` which is a wrapper for the `ruby` command that is invoked.

> But not sure how common that thinking is, that’s why I asked on the mailing list, to hopefully get input from people actually using RVM :-)

I **am** using RVM with TextMate 😉.

> Use `rvm do . which «executable»` to check if the executable was installed for the given Ruby version.

The path comes before the `do` action: `rvm . do which «executable»`

> Did I miss something?

You need to check for the RVM file in the project directory. If there is no such file the above RVM command will fail.

[1] https://rvm.io/integration/textmate

noniq commented:
Great, thanks! (I was under the impression that you’re using rbenv, no idea how I got this mixed up, sorry). Will update the PR accordingly this evening.

noniq commented:
@jacob-carlborg I added RVM support, but I could test it only superficially. Does it work for you?

jacob-carlborg commented on Support/lib/executable.rb:
```diff
@@ -0,0 +1,86 @@
+      elsif File.exist?('Gemfile.lock') && File.read('Gemfile.lock') =~ /^    #{name} /
+        %W(bundle exec #{name})
+
+      elsif (rvm = project_uses_rvm?) && system(*%W(#{rvm} . do which -s #{name}))
```
RVM should not be a fifth alternative. It needs to be checked for and used in each of the existing alternatives.

noniq replied:
You’re right of course, sigh 😒 (I now start to understand why Sam Stephenson was so unhappy with RVM that he decided to create rbenv … )

jacob-carlborg commented on Support/lib/executable.rb:
```diff
@@ -0,0 +1,86 @@
+    # installed and the current dir contains an ini file recognized by rvm, nil
+    # otherwise.
+    def project_uses_rvm?
+      ini_files = %w(.rvmrc .versions.conf .ruby-version .rbfu-version .rbenv-version)
```
`Gemfile` can be used as RVM project file as well [1]. Seems like the official name is "project file".

[1] https://rvm.io/workflow/projects#project-file-gemfile

noniq replied:
Good catch, thanks.

jacob-carlborg commented:
Why `attic` in the `Support/test` directory?

noniq commented:
> Why attic in the Support/test directory?

Couldn’t come up with a better name – open for suggestions :-)

jacob-carlborg commented:
> Couldn’t come up with a better name – open for suggestions :-)

Integration?

noniq commented:
“Integration” sounds like this folder would contain integration tests. But instead it contains test data structures – maybe we should call it “fixtures”?

sanssecours commented:
> But instead it contains test data structures – maybe we should call it “fixtures”?

How about `data` then? This would leave us with the path `test/data`, which sounds about right, according to your description.

By the way: I think you should use the directory name `Tests` instead of `test` (Source: [Bundle Style Guide](http://wiki.macromates.com/Bundles/StyleGuide) – Section “Test Files”).

jacob-carlborg commented:
> “Integration” sounds like this folder would contain integration tests. But instead it contains test data structures – maybe we should call it “fixtures”?

Ah, the test itself is not located in the directory. My mistake. I think `data` sounds good.

jacob-carlborg commented:
> By the way: I think you should use the directory Tests instead of Support/test (Source: Bundle Style Guide – Section “Test Files”). 

Please no. I see the structure in the `Support` directory as the structure of a Ruby gem. The conventsions for gems is to have a directory called `test` (or `spec` if RSpec is being used).

noniq commented:
Updated again. Regarding directory naming and structure I decided to go with `fixtures` and keep the `tests` directory under `Support` (as @jacob-carlborg mentioned, this better mirrors the structure of most Ruby libs).

RVM support is now baked into all the “normal” cases (except environment vars, see commit message), and RVM project file detection should work more reliable now.

sanssecours commented:
> Regarding directory naming and structure I decided to go with fixtures and keep the tests directory under Support (as @jacob-carlborg mentioned, this better mirrors the structure of most Ruby libs).

Sounds good, although I still think the name `fixture` is not that great. Here is what Dictionary says about the term:

>  1. a piece of equipment or furniture that is fixed in position in a building or vehicle: a light fixture.
>     - **(fixtures)** articles attached to a house or land and considered legally part of it so that they normally remain in place when an owner moves: the hotel retains many original **fixtures and fittings**. Compare with fitting ( sense 1 of the noun)).
>     - *informal* a person or thing that is established in a particular place or situation: palm readers were a fixture in most '40s nightclubs.
>  2. *British* a sports event that takes place on a particular date.

None of the above sounds like a good description of test data to me.

jacob-carlborg commented on Support/lib/executable.rb:
```diff
@@ -0,0 +1,93 @@
+    # and the current directory contains an RVM project file.
+    def determine_rvm_prefix
+      rvm = "#{ENV['HOME']}/.rvm/bin/rvm"
+      %W(#{rvm} . do) if File.exist?(rvm) && `#{rvm.shellescape} current`.chomp != 'system'
```
I don't think this will work. It's possible to set a default Ruby using RVM. I think if that is used, the default version will be reported event though there is no RVM project file. That's at least what's happening when I run `"$HOME/.rvm/bin/rvm" current` from within TextMate as a shell script. `rvm . do` always requires a project file, for some reason it doesn't fallback to the default or system Ruby when no project file is found, see https://github.com/rvm/rvm/issues/3746.

noniq replied:
Ah, bonkers, that’s a show stopper of course. Well, let’s wait a bit to see if a solution comes up in the issue you mentioned.

Apart from this, do you think the rest of the implementation makes sense / works correctly?

jacob-carlborg replied:
> Ah, bonkers, that’s a show stopper of course. Well, let’s wait a bit to see if a solution comes up in the issue you mentioned.

I don't think it's a showstopper. You can look for the files as you did before and if there's a Gemfile, read that and look for a line with `^\s*ruby` or similar.

> Apart from this, do you think the rest of the implementation makes sense / works correctly?

I have to look at bit more.

noniq replied:
No activity in the RVM issue until now, so I suppose we just have to bite the bullet and implement a check for a hardcoded list of RVM project files (including the Gemfile). I really don’t like this solution, though … 😬 

Will update the PR shortly.

noniq commented:
I think “test fixtures” is a fairly common term at least in the Rails world: 

> **What Are Fixtures?** Fixtures is a fancy word for sample data. 

(Source: http://guides.rubyonrails.org/testing.html#the-low-down-on-fixtures)

The guide talks about test data in the database only, but Rails itself keeps a lot of file based test data (eg. view templates) in the fixtures directory, too. See for example https://github.com/rails/rails/tree/master/actionpack/test/fixtures

sanssecours commented on Support/lib/executable.rb:
```diff
@@ -0,0 +1,93 @@
+    #       system(*executable, '--some', '--additional', 'args')
+    #     rescue Executable::NotFound => e
+    #       # Executable not found, so use fallback / display alert / …
+    #       # `e.message` contains detailled error message.
```
This sentence contains a small typo:

> …contains **detailled** error message.

Most probably you want

> …contains **detailed** error message.

instead.

noniq replied:
Thx, fixed in 0ec1709! (That word is an old foe of mine, because in German it is spelled “detailliert” with two L :)

jacob-carlborg commented on Support/lib/executable.rb:
```diff
@@ -0,0 +1,93 @@
+    def find(name, env_var = nil)
+      # Safeguard against invalid names so that we don’t need to care about
+      # shell escaping later on
+      raise ArgumentError, "Invalid characters found in '#{name}'" unless name =~ /\A[\w_-]+\z/
```
Why not use `Shellwords.shellescape` instead?

noniq replied:
I see this more like some kind of input validation: An executable name should not contain any characters that need to be shellescaped in the first place. Are there any possible use cases where this validation is too strict?

jacob-carlborg replied:
No, not that I can think of. But I think it's cleaner to use the built-in method to handle this. It's also simpler and doesn't add any artificial limitations. As I said before, we should be very careful making assumptions on which commands might be invoked.

noniq replied:
I like to be conservative and explicitly forbid use cases I have not thought of (because haven’t thought of a use case also means that there are no tests for it 😄). But I’ll be happy to relax this check if an actual use case shows up.

jacob-carlborg commented on Support/lib/executable.rb:
```diff
@@ -0,0 +1,93 @@
+
+      env_var ||= 'TM_' + name.gsub(/\W+/, '_').upcase
+      prefix = determine_rvm_prefix || []
+      if (cmd = ENV[env_var]) && cmd != ''
```
I'm not a big fan of declaring variables in if statements in Ruby, because the scope will be outside of the if statement.

noniq replied:
I generally agree, but I think in this case it’s justifiable to keep the code concise.

noniq commented:
So let’s try to get this shipped! 🚢 

I reverted RVM project file detection back to the old method, and also added checking for a Ruby version in the Gemfile. Is there anything else left?

jacob-carlborg commented:
I'll give it another look.

jacob-carlborg commented on Support/lib/executable.rb:
```diff
@@ -0,0 +1,106 @@
+      prefix = determine_rvm_prefix || []
+      if (cmd = ENV[env_var]) && cmd != ''
+        cmd = cmd.shellsplit
+        if system('which', '-s', cmd[0])
```
What happens if you put something like `bundle exec foo` or `rvm . do foo` in an environment variable?

noniq replied:
Should work, there is even a test for it: https://github.com/die-antwort/ruby.tmbundle/blob/implement-find-executable/Support/test/test_executable.rb#L79-L84 :-)

jacob-carlborg commented on Support/lib/executable.rb:
```diff
@@ -0,0 +1,106 @@
+      elsif File.exist?('Gemfile.lock') && File.read('Gemfile.lock') =~ /^    #{name} /
+        prefix + %W(bundle exec #{name})
+
+      elsif (path = `which #{name}`.chomp) != ''
```
Doesn't this need to take RVM into consideration.

noniq replied:
That’s a good point, hmm … Not sure right know, I’ll think about it a bit more.

jacob-carlborg commented on Support/lib/executable.rb:
```diff
@@ -0,0 +1,106 @@
+    end
+
+    def ruby_version_in_gemfile?
+      File.exist?('Gemfile.lock') && File.read('Gemfile.lock') =~ /^RUBY_VERSION/
```
This might be an edge case, but if one implements a command to run `bundle install` and uses this method to find `bundle` the lock file might not exist yet.

noniq replied:
Right, but as you say this is an edge case and I don’t think it’s worth bringing additional complexity into the code just to deal with this.

jacob-carlborg replied:
I was not thinking about bringing in more complexity, I was thinking about reading `Gemfile` instead of `Gemfile.lock`.

noniq replied:
`Gemfile` is harder to parse (it’s basically plain Ruby code), wherease the `Gemfile.lock` has a fixed format. (Also, we already require `Gemfile.lock` to be present when searching for executables in the gems.)

noniq commented:
Okay, so one more thought about RVM (caused by https://github.com/textmate/ruby.tmbundle/pull/104#discussion_r81365193): 

Currently, we use RVM only if we find a RVM project file. This has two problems:
 * It is rather complicated (RVM supports many different project files, as well as simply specifying a ruby version in `Gemfile`:
 * It won’t work (I suppose) if you have a default Ruby version set via RVM (and no project file in the current project).

So what about simply removing project file detection and using RVM always if it installed? (That would also mirror the behaviour for rbenv: If it is installed, use it).

jacob-carlborg commented:
> So what about simply removing project file detection and using RVM always if it installed? (That would also mirror the behaviour for rbenv: If it is installed, use it).

How would that solve your second point?

noniq commented:
Right … I initially thought of simply always prefixing commands with `rvm . do` – but as you already mentioned, this won’t work in the case a default Ruby is set. Okay, so for detecting an executable in path we also have to check if RVM is installed and a project file is present, and if so, use `rvm . do which …`. Will update the code accordingly.

noniq commented:
Okay, should be fixed now with 9e4b970!

jacob-carlborg commented:
An alternative way to deal with this RVM stuff could be to just invoke `rvm . do` with a command that should always pass, something like `rvm . do true`. If that command fails it means that no project file exists.

BTW, I noticed that invoking a command using `rvm . do` adds significant overhead. Invoking `true` takes 0.05 seconds on my machine, will invoking the same command prefixed with `rvm . do` takes up to 0.498 seconds.

noniq commented:
That’s a nice idea, using `rvm . do true`! I’m not sure what to do about the overhead, though … any ideas?

jacob-carlborg commented:
> I’m not sure what to do about the overhead, though … any ideas?

No, not really. Provide an API that will run the command in the background and recommend using that in the documentation?

noniq commented:
You could use `TextMate.detach` with a block to then run the executable found by `Executable.find` in the background. But it still means that we go through the overhead twice – once when finding the executable, and (at least) once more when it is actually invoked. Not sure we can do anything about that 😞

noniq commented:
I have an idea regarding the overhead when detecting a RVM executable: 

We could introduce some kind of power user setting like `TM_ALWAYS_ASSUME_RVM`, and if this is set we’d just skip the check in [line 69](https://github.com/textmate/ruby.tmbundle/pull/104/files#diff-c928c153bcb20fd44cf2fe76e200d9a6R69) and simply return `rvm . do <executable>`.

This would save the overhead of invoking `rvm . do` just to check if the executable is installed, for the cost of harder to debug errors if an expected executable is indeed not installed. But users using this setting this would be expected to be able to deal with that.

jacob-carlborg commented:
I'm not sure. @sorbits @infininight ideas?

noniq commented:
Any news on this? Shall we proceed and get this merged?

jacob-carlborg commented:
@noniq what do you think about this: https://github.com/rvm/rvm/issues/3746#issuecomment-255589461 ?

noniq commented:
I think this would work only if the command is run via a subshell that executes the rvm initialization function, wouldn’t it? And wouldn’t that have the same overhead as running `rvm . do which …`?

jacob-carlborg commented:
> I think this would work only if the command is run via a subshell that executes the rvm initialization function, wouldn’t it

Yes.

> And wouldn’t that have the same overhead as running rvm . do which …?

Not sure. Sourcing RVM takes 18ms, but navigating to a directory with a project file takes around 140ms.

But we would not need to manually check for the project files.

noniq commented:
Back at this after 2 busy weeks: Regarding RVM, if we want to run the command via a subshell I think we’ll need to provide a tiny wrapper script (that does nothing more than `cd $directory && $executable`). This is because I want `#find_executable` to always return an array (makes it easier to append additional arguments, doesn’t need to spawn a subshell when used with `system`, `exec`, `popen` etc.)

Should I give this a try? What would be the definitive way to initialize rvm in this wrapper script?

jacob-carlborg commented:
> What would be the definitive way to initialize rvm in this wrapper script?

RVM adds the following to `.bashrc` or similar:

```bash
export PATH="$PATH:$HOME/.rvm/bin" # Add RVM to PATH for scripting

[[ -s "$HOME/.rvm/scripts/rvm" ]] && source "$HOME/.rvm/scripts/rvm" # Load RVM into a shell session *as a function*
```

I think that the last line is all that's needed.

noniq commented:
Thanks for the quick response!

One more question: Will this work correctly if the working directory is alread set to the project directory? I.e. does `cd .` work correctly with rvm? Or do we need some trickery like `cd $(pwd)`?

noniq commented:
I just implemented this (currently using `cd .` in the wrapper script). @jacob-carlborg: Please let me know if this works for you!

noniq commented:
Any news on this ( @jacob-carlborg, @sanssecours)?

(There’s a recent issue from my RuboCop bundle concerning RVM: https://github.com/noniq/RuboCop.tmbundle/issues/1 – would be perfect if I could fix that issue by making the RuboCop bundle use `Executable.find`)

sanssecours commented:
> Any news on this ( @jacob-carlborg, @sanssecours)?

Basically I am just [waiting](https://github.com/textmate/ruby.tmbundle/pull/99) for this pull request to get through, so I can add the [Ruby version](https://github.com/sanssecours/ruby.tmbundle/commit/4c73c8e97254601118472f203a298fe11e1cb9ff) of “Reformat Code“. 

The master branch of my clone of the Ruby bundle already uses the latest version of `Executable.find`. In my experience – with “Reformat Document” – so far `Executable.find` works just fine. However, as I already mentioned before, I do not use RVM at all.

> There’s a recent issue from my RuboCop bundle…

Oh, another RuboCop bundle 😀. I think I will try that one too…

sanssecours commented:
> Any news on this ( @jacob-carlborg, @sanssecours)?

Basically I am just waiting for this pull request to get trough, so I can add the [Ruby version](https://github.com/sanssecours/ruby.tmbundle/commit/4c73c8e97254601118472f203a298fe11e1cb9ff) of “Reformat Document” to my [pull request](https://github.com/textmate/ruby.tmbundle/pull/99).

The master branch of my clone of the Ruby bundle already uses the latest version of `Executable.find`. From my experience – with “Reformat Document” – so far, `Executable.find` works just fine. However, as I already mentioned before, I do not use RVM at all.

> There’s a recent issue from my RuboCop bundle

Oh, another RuboCop bundle 😀. I think I will try that one too…

jacob-carlborg commented:
Overall I think it looks fine. I'll try tomorrow to actually test it to see that it works properly.

jacob-carlborg commented:
Most things seem to work, although I noticed two things that don't work:

1. It's not possible to use a full path since `/` is not allowed in the regular expression
2. `.` is not allowed in the regular expression

noniq commented:
The purpose of `Executable.find` is to find a ruby executable when you know the name only. If you already know / have a path, you don’t need this function at all, so not allowing `/` in the `name` argument is on purpose. 

Ad 2.: Are there any valid Ruby executables with a `.` in their name? I can’t think of any, but I’ll happily relax add this to the regex if I’m proven wrong :-)

jacob-carlborg commented:
> The purpose of Executable.find is to find a Ruby executable when you know the name only. If you already know / have a path, you don’t need this function at all, so not allowing / in the name argument is on purpose.

It would handle RVM automatically.

> Ad 2.: Are there any valid Ruby executables with a . in their name? I can’t think of any, but I’ll happily extend the regex if I’m proven wrong :-)

I don't know, but I tried just executing a Ruby file (with the `.rb` extension) and that failed.

I don't understand why you're insisting on adding artificial limitations.

noniq commented:
I want this method to serve a precisely defined use-case: Finding ruby-installed executables like rspec, rubocop or kramdown. Nothing less, but nothing more neither (the code is already to complex for my taste, but I haven’t yet found a way to simplify it while keeping all the features that are necessary).

noniq commented:
Remember that the original idea behind this PR was to provide this functionality as a service to other bundles. There are several bundles (eg. RSpec, various RuboCup ones, the Ruby bundle itself [see #99]) that would profit from using `Executable.find` in exact the way it is now designed to be used. 

If additional use cases show up in the future, we can of course change or extend this feature at any time. But for now I think we should get this finally merged, so that other bundles can start using it. Then we’ll see how it goes. 😀

jacob-carlborg commented:
Then I think this looks good 👌. Unfortunately I don't have permission to merge. @sorbits @infininight please have a look.

noniq commented:
Gentle ping @sorbits @infininight :) 

TL;DR: I don’t think you need to re-read all of the discussion here in detail. See #102 for the rationale behind the feature and [this comment](https://github.com/die-antwort/ruby.tmbundle/blob/149498e129c3a29d7ba3776eba208524921e1402/Support/lib/executable.rb#L7-L47) for an explanation of how it works and how to use it.

Most of the (loooooong) discussion here was around specific details of how to make this feature work with all kinds of setups people might use nowadays (rvm, rbenv, bundler, …). In the end we arrived at a nice (and [well-tested](/textmate/ruby.tmbundle/blob/149498e129c3a29d7ba3776eba208524921e1402/Support/test/test_executable.rb)) implementation.

I’m rather sure that there still are some cases we haven’t covered, but these will only start to turn up when this feature starts getting used. I’ll happily commit to continue supporting this feature in the future (as long as I’m still using TextMate, but no plans to change that for now 🙃).

noniq commented:
Ping @infininight @sorbits (see above for the TL;DR 🙂)

sorbits commented:
Sorry about the delay, I have squashed all the commits into f49a2f5113745a0c88c29924329eb1ca896d0f9e and merged that, I hope this is acceptable.

noniq commented:
Great, thanks!
